### PR TITLE
Add EIP: NONREENTRANT and REENTRANT opcodes

### PR DESCRIPTION
**Pull request by charles-cooper**
**ATTENTION: ERC-RELATED PULL REQUESTS NOW OCCUR IN [ETHEREUM/ERCS](https://github.com/ethereum/ercs)**

--

When opening a pull request to submit a new EIP, please use the suggested template: https://github.com/ethereum/EIPs/blob/master/eip-template.md

We have a GitHub bot that automatically merges some PRs. It will merge yours immediately if certain criteria are met:

 - The PR edits only existing draft PRs.
 - The build passes.
 - Your GitHub username or email address is listed in the 'author' header of all affected PRs, inside <triangular brackets>.
 - If matching on email address, the email address is the one publicly listed on your GitHub profile.

### CODE REVIEW

File: EIPS/nonreentrant.md [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+created: 2024-05-09
+---
+
+## Abstract
+
+Add two opcodes, `NONREENTRANT` and `REENTRANT`, which set and clear a contract's reentrancy status. After invoking `NONREENTRANT`, a contract cannot be `CALL`ed (or `STATICCALL`ed, or `DELEGATECALL`ed) until `REENTRANT` is invoked.
+
+## Motivation
+
+Reentrancy attacks account for a substantial portion of user funds stolen on EVM chains, including the famous "DAO hack" (a more complete list is available at https://github.com/pcaversaccio/reentrancy-attacks). However, due to the cost of preventing against reentrancy attacks in application code, developers often opt-out of reentrancy protection. This cost has come down with the advent of transient storage (EIP-1155), but it is still not cheap enough where it is a "no-brainer" to use it by default. This EIP proposes opcodes which make it cheaper to protect against reentrancy in application code.
```
Comment (pcaversaccio) [inline EIPS/nonreentrant.md:18]:
```suggestion
Reentrancy attacks account for a substantial portion of user funds stolen on EVM chains, including the famous "DAO hack" (a more complete list is available at https://github.com/pcaversaccio/reentrancy-attacks). However, due to the cost of preventing against reentrancy attacks in application code, developers often opt-out of reentrancy protection. This cost has come down with the advent of transient storage ([EIP-1153](./eip-1153.md)), but it is still not cheap enough where it is a "no-brainer" to use it by default. This EIP proposes opcodes which make it cheaper to protect against reentrancy in application code.
```

File: EIPS/nonreentrant.md [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+
+## Specification
+
+The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED", "NOT RECOMMENDED", "MAY", and "OPTIONAL" in this document are to be interpreted as described in RFC 2119 and RFC 8174.
+
+Two new opcodes are introduced, `NONREENTRANT` (0xF6) and `REENTRANT` (0xF7), which set and clear a contract's nonreentrancy flag. The effect of invoking `NONREENTRANT` is that a contract can no longer have execution context transferred to it (via any of the `*CALL` opcodes), until `REENTRANT` is invoked. `CALL`ing a contract which has the nonreentrancy flag set is equivalent to executing a single `REVERT` opcode.
```
Comment (pcaversaccio) [inline EIPS/nonreentrant.md:24]: FYI, `0xf6` is `AUTH` and `0xF7` is `AUTHCALL` if EIP-3074 makes it...

File: EIPS/nonreentrant.md [hunk 3]
```diff
@@ -0,0 +1,52 @@
+---
```
Comment (abcoathup) [inline EIPS/nonreentrant.md:1]:
```suggestion
--- 
eip: 7705
```
Assigning next sequential EIP/ERC/RIP number.

Please also update the filename.

File: EIPS/nonreentrant.md [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+title: NONREENTRANT and REENTRANT opcodes
+description: Opcodes to mark a contract as nonreentrant
+author: Charles Cooper (@charles-cooper)
+discussions-to: https://ethereum-magicians.org/t/eip-for-nonreentrant-opcodes/19957
```
Comment (abcoathup) [inline EIPS/nonreentrant.md:5]:
```suggestion
discussions-to: https://ethereum-magicians.org/t/eip-7705-nonreentrant-opcodes/19957
```

**Review by mouhand999 — APPROVED** [review]

**Review by g11tech — APPROVED** [review]
lgtm

**Pull request merged**